### PR TITLE
Association generator

### PR DESCRIPTION
**Pull request by SamirBoulil**
- Contains fixes to work with pim 1.5
- Adds an association generator
- Backports .php_cs-fixers.php .php_cs.php .scrutinizer.yml .travis.yml from master

The association generator command works as a standalone command that is executed after a product import.
It is able to generate product associations as well as group associations.

| Q | A |
| --- | --- |
| Added Specs | - |
| Generate fixtures and install | yes |
| Import generated products | yes |
| Import generated associations | yes |
| Review and 2 GTM | no |
|  Tech Doc | yes (see README) |

### CODE REVIEW

File: Generator/AssociationGenerator.php [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        if (0 === $this->groupRepository->countVariantGroups() && $associationsVariantGroupCount > 0) {
+            throw new \LogicException('No variant groups found to generate group associations');
+        }
+
+        $this->faker = Factory::create();
+        if (isset($globalConfig['seed'])) {
+            $this->faker->seed($globalConfig['seed']);
+        }
+
+        foreach ($products as $i => $product) {
```
Comment (pierallard) [inline Generator/AssociationGenerator.php]: Note for later (when we will have a cs fixer) : don't use one-letter variable.

File: Generator/AssociationGenerator.php [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+                $productIdentifiers[] = $product->getIdentifier()->getData();
+                $i++;
+            }
+
+            $products->next();
+
+            $this->objectDetacher->detach($product);
+        }
+
+        return join(',', $productIdentifiers);
```
Comment (pierallard) [inline Generator/AssociationGenerator.php]: please move this join, rename this method into generateRandomProductIdentifiers, and insert the join into line 113

File: Generator/AssociationGenerator.php [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    /**
+     * Find product skus different from the product passed as parameter
+     *
+     * @param ProductInterface $excludedProduct
+     * @param CursorInterface  $products
+     * @param int              $associationCount
+     *
+     * @return array
```
Comment (pierallard) [inline Generator/AssociationGenerator.php:163]: Yes, please, return an array ;)

File: Generator/AssociationGenerator.php [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        foreach ($products as $i => $product) {
+            if (null !== $associationLimitNumber && $i >= $associationLimitNumber) {
+                break;
+            }
+
+            $association = [];
+            $association[$identifierCode] = $product->getIdentifier()->getData();
+
+            $associationTypeProduct = $this->getRandomAssociations('-products');
+            $randomProducts = $this->generateRandomProducts(
```
Comment (pierallard) [inline Generator/AssociationGenerator.php]: randomProductIdentifiers

File: Generator/AssociationGenerator.php [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        $variantGroups = $this->faker->randomElements(
+            $this->groupRepository->getAllVariantGroups(),
+            $associationsVariantGroupCount
+        );
+
+        $variantGroupCodes = array_map(function ($variantGroup) {
+            return $variantGroup->getCode();
+        }, $variantGroups);
+
+        return join(',', $variantGroupCodes);
```
Comment (pierallard) [inline Generator/AssociationGenerator.php]: same here (with update phpdoc)

File: Generator/AssociationGenerator.php [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+    /**
+     * @return CursorInterface
+     */
+    private function getProducts()
+    {
+        $productQueryBuilder = $this->productQueryBuilderFactory->create();
+        $products = $productQueryBuilder->execute();
+
+        return $products;
```
Comment (pierallard) [inline Generator/AssociationGenerator.php:224]: whow, this does not explode when there is 3000000 products ?

**Comment by pierallard** [comment]
GG maggle. Some comments, then GTM :heart:
Thx samir

File: Generator/AssociationGenerator.php (hunk 1 above)
Comment (SamirBoulil) [inline Generator/AssociationGenerator.php]: done

File: Generator/AssociationGenerator.php (hunk 2 above)
Comment (SamirBoulil) [inline Generator/AssociationGenerator.php]: done

File: Generator/AssociationGenerator.php (hunk 3 above)
Comment (SamirBoulil) [inline Generator/AssociationGenerator.php:163]: not done :trollface:

File: Generator/AssociationGenerator.php (hunk 4 above)
Comment (SamirBoulil) [inline Generator/AssociationGenerator.php]: generateRandomProductIdentifiers instead. Done

File: Generator/AssociationGenerator.php (hunk 5 above)
Comment (SamirBoulil) [inline Generator/AssociationGenerator.php]: Done

File: Generator/AssociationGenerator.php (hunk 6 above)
Comment (SamirBoulil) [inline Generator/AssociationGenerator.php:224]: Why would it ? We use the productQueryBuilder that batches the queries (by sets of 100).
Thats what we used everywhere in the PIM to iterate over products.

Do you have any case in my mind the pqb didn't work ?

Comment (pierallard) [inline Generator/AssociationGenerator.php:224]: No no, I thought it returns all the products of the PIM.

**Pull request merged**